tf_quant_finance/black_scholes/brownian_bridge.py

```diff
@@ -0,0 +1,130 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow.compat.v2 as tf
+
+
+def brownian_bridge_double(*,
+                           x_start,
+                           x_end,
+                           variance,
+                           upper_barrier,
+                           lower_barrier,
+                           n_cutoff=3,
+                           dtype=None,
+                           name=None):
+  """ Probability of not touching the upper barrier and lower_barrier
+  for a Brownian Bridge starting at `x_start`, ending at `x_end` and with
+  variance `variance`, assuming that `x_start` and `x_end` are within the
+  barriers.
+  This can be used in Monte Carlo pricing for adjusting probability of
+  touching the barriers from discrete case to continuous case.
+  Typically in practise, the tensors `x_start`, `x_end` and `variance` should be
+  bi-dimensional (with time steps and paths being the 2 dimensions).
+
+  ## References
+
+  [1] Emmanuel Gobet. Advanced Monte Carlo methods for barrier and related
+  exotic options.
+  https://papers.ssrn.com/sol3/papers.cfm?abstract_id=1265669
+
+  Args:
+    x_start: Real `Tensor` of any shape and dtype
+    x_end: A real `Tensor` of the same dtype and compatible shape as
+      `x_start`
+    variance: A real `Tensor` of the same dtype and compatible shape as
+      `x_start`
+    upper_barrier: A scalar `Tensor`
+    lower_barrier: A scalar `Tensor`
+    n_cutoff: A positive scalar int32 `Tensor`. This controls when to cutoff
+      the sum which would otherwise have an infinite number of terms.
+      Default value: 3.
+    dtype: Optional `tf.DType`. If supplied, the dtype to be used for conversion
+      of any supplied non-`Tensor` arguments to `Tensor`.
+      Default value: None which maps to the default dtype inferred by TensorFlow
+        (float32).
+    name: str. The name for the ops created by this function.
+      Default value: None which is mapped to the default name `option_price`.
+
+  Returns:
+      A `Tensor` of the same shape as the input data which is the probability
+      of not touching the upper and lower barrier.
+  """
+  with tf.name_scope(name or 'double_brownian_bridge'):
+    x_start = tf.convert_to_tensor(x_start, dtype=dtype, name='x_start')
+    dtype = x_start.dtype
+    variance = tf.convert_to_tensor(variance, dtype=dtype, name='variance')
+    x_end = tf.convert_to_tensor(x_end, dtype=dtype, name='x_end')
+    barrier_diff = upper_barrier - lower_barrier
+
+    def summed_func(k):
+      return tf.exp(-2 * k * barrier_diff * (
```

Comment:
`tf.math.exp` for consistency (since, e.g., tf.log is not exposed and we use tf.math.log instead)